websockets/websocket-next/src/test/java/io/quarkuss/ts/websocketNext/BaseWebSocketIT.java

```diff
@@ -0,0 +1,377 @@
+         * In concurrent mode, the later message is faster and should arrive first.
+         */
+
+        Client serialClient = createClient("/serial");
+        Client concurentClient = createClient("/concurrent");
+
+        serialClient.send("block");
+        serialClient.send("go");
+        assertMessage("block", serialClient);
+        assertMessage("go", serialClient);
+
+        concurentClient.send("block");
+        concurentClient.send("go");
+        assertMessage("go", concurentClient);
+        assertMessage("block", concurentClient);
+    }
+
+    @Test
+    public void userDataOnServerTest() throws URISyntaxException, InterruptedException {
+        Client client = createClient("/userData");
+
+        client.send("get");
+        assertMessage("messages sent: 0", client);
+
+        client.send("one");
+        assertMessage("one", client);
+        client.send("two");
+        assertMessage("two", client);
+        client.send("three");
+        assertMessage("three", client);
+
+        client.send("get");
+        assertMessage("messages sent: 3", client);
+    }
+
+    @Test
+    public void userDataInClientTest() throws URISyntaxException, InterruptedException {
+        Client client = createClient("/chat/alice");
+        // connect server-side client
+        given().queryParam("username", "bob").get("/userDataRes/connect");
+
+        try {
+            // server-side client will respond to "login" message by sending its username (which is stored in user data)
+            client.send("login");
+            assertMessage("alice: login", client);
+            // first username is set by the chat broker, second one by the server-side client
+            assertMessage("bob: bob", client);
+        } finally {
+            given().get("/userDataRes/disconnect");
+        }
+    }
+
+    @Test
+    public void subWebSocketTest() throws URISyntaxException, InterruptedException {
+        Client parentClient = createClient("/parent");
+        Client nestedClient = createClient("/parent/nested");
+
+        parentClient.send("foo");
+        assertMessage("This is parent webSocket", parentClient);
+
+        nestedClient.send("bar");
+        assertMessage("This is nested webSocket", nestedClient);
+    }
+
+    // verify that custom logic for allowing/rejecting upgrade http-to-websocket works
+    @Test
+    public void httpToWebSocketUpgradeTest() throws URISyntaxException, InterruptedException {
+        Client rejectClient = new Client(getUri("/parent"), true);
+        rejectClient.addHeader("Reject", "");
+        assertFalse(rejectClient.connectBlocking(), "Upgrade from http to websocket should be rejected");
+
+        Client allowClient = new Client(getUri("/parent"), true);
+        assertTrue(allowClient.connectBlocking(), "Upgrade from http to websocket should be allowed");
+    }
+
+    @Test
+    public void pingPongTest() throws InterruptedException {
```

Comment:
This whole test is based on data race. Please, consider other options

Reply:
I don't really see many other options how to test that if I configure client to send every 2 seconds, that it actually sends it every two seconds. I can change the numbers so there is less room for flukes. Or I can change it await() until some pings appear and then validate time it took, but I'm not sure if that is much better. WDYT @fedinskiy ?

Reply:
I've changed the test, to verify time difference between ping pong events.